src/presign/participant.rs

```diff
@@ -455,29 +446,32 @@ impl PresignParticipant {
             &r2_priv_ij,
         )?;
 
-        let out_message = Message::new(
+        let round_one_outcome = ProcessOutcome::Processed(vec![Message::new(
             MessageType::Presign(PresignMessageType::RoundTwo),
             message.id(),
             self.id,
             message.from(), // This is a essentially response to that sender
             &serialize!(&r2_pub_ij)?,
-        );
+        )]);
 
-        let mut messages = vec![out_message];
         // Check if there's a round 2 message that this now allows us to process
         let retrieved_messages = self.fetch_messages_by_sender(
             MessageType::Presign(PresignMessageType::RoundTwo),
             message.id(),
             message.from(),
         )?;
-        let r2_message = match retrieved_messages.get(0) {
-            Some(message) => message,
-            None => return Ok((None, messages)),
-        };
-        let (presign_record_option, mut additional_messages) =
-            self.handle_round_two_msg(rng, r2_message, main_storage)?;
-        messages.append(&mut additional_messages);
-        Ok((presign_record_option, messages))
+
+        let round_two_outcomes = retrieved_messages
+            .iter()
+            .map(|msg| self.handle_round_two_msg(rng, msg, main_storage))
+            .collect::<Result<Vec<_>>>()?;
+
+        if round_two_outcomes.len() > 1 {
+            // There should never be more than one round 2 message from a single party
+            Err(InternalError::InternalInvariantFailed)
```

Comment:
I just want to double check that this should be an `InternalInvariantFailed`. Suppose a malicious party sends multiple round 2 messages, is the idea that this will be caught _before_ we reach this point?

Reply:
That's a good point. I'll do a protocol error instead and add some logging.

I'm not sure if we actually monitor this case in general (e.g. where a malicious party sends multiple versions of a message for a round). Something to consider in testing eventually.